[ananya21] iP

ananya21 opened the pull request:
# Project Charlie Enhancement 🚀

> "Innovation distinguishes between a leader and a follower." - Steve Jobs

This PR introduces several enhancements to the **Project Charlie**, improving its performance and user experience. Here's what's new:

## Features:

- **Enhanced Speed**:
  - Now **_50% faster_** than before!
- **New User Interface**:
  - A sleek, modern UI for an improved user experience.

## How to Get Started:

1. **Download** the latest version from [here](https://google.com).
2. **Install** it by double-clicking the downloaded file.
3. **Enjoy** the new and improved features!

## What's Included in This PR:

```java
public class Enhancements {
    public static void main(String[] args) {
        System.out.println("New Features Implemented!");
    }
}

ananya21 committed:
Add Level-0: Basic Greet and Exit Functionality

ananya21 committed:
Add Level-1: Echo

ananya21 committed:
Add Level-2: Adding to a List

ananya21 committed:
Add Level-3: Marking Tasks

ananya21 committed:
Completed Level-4 with times

ananya21 committed:
Added automated testing

ananya21 committed:
Handled errors

wilsonwid commented on src/main/java/Charlie.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
```
It would be good to add a Javadoc comment for the class, as well as the public methods inside the class.

wilsonwid commented on src/main/java/Charlie.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.ArrayList;
```
It would be good to place all files in a package.

wilsonwid commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) {
```
It would be good to add Javadoc comments to public methods and to the class.

wilsonwid commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+        return (isDone ? "[X]" : "[ ]"); // mark done task with X
+    }
+
+    public void markAsDone() {
```
Could this method and `markAsNotDone()` perhaps be fused into one method with a `boolean` parameter?

wilsonwid commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,11 @@
+    }
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
```
I like how inheritance and polymorphism are used to shorten this line.

wilsonwid reviewed (commented):
The code in this PR is of an overall good standard, even though there are several places where it could be improved. I commented on several coding standard violations, most of which are due to missing Javadoc comments. It would also be good to review the imports used in each of the classes to see whether they are unused.

NusMinato reviewed (commented):
No code naming issues found. Well done!

ananya21 committed:
Add Gradle Support

ananya21 committed:
Add Gradle Support

ananya21 committed:
Added JUnit tests

ananya21 committed:
Added JavaDoc

ananya21 committed:
Adhered to coding standard

ananya21 committed:
Find Level-9

ananya21 committed:
Find Level-9

ananya21 committed:
Added JAR file

ananya21 committed:
Used checkstyle to fix some coding violations

ananya21 committed:
implemented GUI

ananya21 committed:
Task class: add assertions to validate state

Ensure Task objects are in a consistent state

In our Task class, certain assumptions about the state of Task objects
hold true throughout the lifecycle of these objects (e.g., description
cannot be null). Previously, these assumptions were not explicitly
checked, leading to a risk of encountering NullPointerExceptions if
these assumptions ever got violated due to future changes.

Let's,
* add assertions to the constructor to ensure the description is not null
* add assertions in setStatus method to validate the state transitions

Adding these assertions helps catch bugs during the development phase
and ensures our assumptions about the Task object's state are always
validated.

ananya21 committed:
Refactor Parser for Enhanced Code Quality

Extract command initialization into dedicated methods

The Parser class's single-method approach for command parsing was
becoming cumbersome and difficult to maintain. This refactoring aims
to improve the code's readability and maintainability by adhering
to the Single Responsibility Principle.

Let's,
* Create separate methods for each command's parsing logic, such as
  , , etc.
* Ensure each method is responsible for a single aspect of the
  command parsing process.

This change not only makes the Parser class easier to understand
and modify but also simplifies adding new commands in the future.

By breaking down the complex  method into smaller, focused
methods, we improve the code quality significantly, making it
easier for future developers to navigate and extend the application.

ananya21 committed:
Implement task prioritization feature in Parser and Duke classes

Introduce the ability to assign priority levels to tasks, allowing users to
mark tasks with a specific priority. This enhancement enables more effective
task management by emphasizing the urgency or importance of certain tasks over others.

Modifications include:
- Update the Parser class to recognize priority commands and extract the relevant
  priority level from user input.
- Adjust the Duke class to handle the assignment of priority levels to tasks,
  storing this information accordingly.

This change addresses the need for users to organize their tasks by priority,
facilitating a more efficient and targeted approach to task management.

Priority levels are defined as high, medium, and low, corresponding to numeric
values for easy comparison. The implementation supports dynamic
adjustment of priorities, allowing users to update the priority level of tasks
as needed.

ananya21 committed:
Update README with comprehensive feature guide

Enhance the User Guide in the README with detailed descriptions of new features

This update enriches the README file with a comprehensive guide to the application's features, improving user understanding and accessibility. The changes include:

- Detailed instructions on how to add todo, deadline, and event tasks, including how users are prompted for task prioritization.
- Explanation of the find keyword feature for locating tasks related to specific keywords.
- Guidelines on deleting tasks, listing all tasks, marking tasks as completed, and unmarking tasks.
- Usage examples for each command, providing clear scenarios that demonstrate the expected outcomes.

These enhancements aim to provide users with clear and concise instructions on utilizing the application's functionality to its full extent, promoting a more intuitive and user-friendly experience.

Additionally, the inclusion of expected outcomes for each command usage example serves to set clear expectations, aiding users in verifying the correct application behavior.

ananya21 committed:
Adding Charlie to ReadMe file

ananya21 committed:
Fixing bug

ananya21 committed:
Resolved merge conflicts
